File: backend/app/schemas/transfers.py
from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field

from app.services.transfer_types import WalletType


class TransferRequest(BaseModel):
    credential_id: UUID | None = None
    asset: str = Field(min_length=1, max_length=20)
    amount: Decimal = Field(gt=0)
    from_wallet: WalletType
    to_wallet: WalletType


class TransferResponse(BaseModel):
    id: UUID
    asset: str
    amount: Decimal
    from_wallet: str
    to_wallet: str
    binance_tran_id: str | None
    status: str
    error_message: str | None
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/api/transfers.ts
import { apiClient } from "./client";

export type WalletType = "SPOT" | "USDM_FUTURES" | "FUNDING";

export interface Transfer {
  id: string;
  asset: string;
  amount: string;
  from_wallet: WalletType;
  to_wallet: WalletType;
  binance_tran_id: string | null;
  status: "PENDING" | "SUCCESS" | "FAILED";
  error_message: string | null;
  created_at: string;
}

export async function listTransfers(): Promise<Transfer[]> {
  const { data } = await apiClient.get<Transfer[]>("/transfers");
  return data;
}

export async function createTransfer(params: {
  credential_id?: string;
  asset: string;
  amount: string;
  from_wallet: WalletType;
  to_wallet: WalletType;
}): Promise<Transfer> {
  const { data } = await apiClient.post<Transfer>("/transfers", params);
  return data;
}
